### PR TITLE
kubelet: get IP based on service network IP mode for dual-stack support.

### LINKED ISSUE #70653: dual-stack: kubelet preferred family
**What would you like to be added**:
To have kubelet get the pod's IP, based on the family used for the service network. For an IPv4 only cluster, this would be the pod's IPv4 address. For an IPv6 only cluster, this would be the pod's IPv6 address. For dual-stack, this would be the IPv4 address, if the service network is using IPv4, or the IPv6 address, if the service network is using IPv6.

**Why is this needed**:
For dual-stack networking, as being proposed in KEP 2245, the service network will be single family. For proper operation, the pod IP being used should match the family used by the service network.

Currently, kubelet will prioritize using the IPv4 address from a pod, and if it does *not* exist, an attempt will be made to use the IPv6 address from the pod. This works for IPv4 only and IPv6 only clusters, but will not work for dual stack.

I listed this as an enhancement, instead of a bug, as current operation for IPv4 only and IPv6 only work fine. Let me know, if the issue should be recategorized.

I have changes for a PR in progress and will push up as soon as I can.

/sig networking
/area ipv6

/kind feature

### PR DESCRIPTION
**Pull request by pmichali**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Allows kubelet to obtain the correct IP from pods, when operating in dual-stack mode.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #70653

**Special notes for your reviewer**:
Intent of this is to allow IPv4 only and IPv6 only to work as they do today, but adapt kubelet for when running in dual-stack cluster.

**Does this PR introduce a user-facing change?**:
NONE

/area ipv6
/sig network

### CODE REVIEW

**Comment by pmichali** [comment]
Commit message for this change...

    This is WIP as it relies on bug fix 70633 that fixes reggression in etcd. Once
    that commit merges, this can be rebased.

    In GetPodIP, the kubelet code attempts to get the IPv4 address for the
    pod, and if that fails, tries to get the IPv6 address.  That works fine
    for IPv4 only and IPv6 only mode, but not for dual-stack, where each
    pod will have both addresses.

    In addition, since dual-stack will support only a single (selectable)
    family for services, we need to also ensure that kubelet is using the
    same family, when getting the IP. One way to do that is to request
    the IP, based on the family used for services.

    With the introduction of IPv6 only mode, a DNS_SVC_IP environment
    variable was defined, so that DNS used an IP that was in the family
    for services. The variable can be provided to the kubelet, via a
    drop-in file, as is done for kubeadm-dind-cluster.

    This change makes use of that information, reading the variable, and,
    if set, will use the same family as the DNS IP in the request to obtain
    the pod IP. Otherwise, it will fall back on trying to get the IPv4
    address, and then trying to get the IPv6 address, if no IPv4 address
    is available (to support backward compatibility).

**Comment by pmichali** [comment]
NOTE: this is WIP because regression was found in etcd for IPv6 clusters and this commit includes the cherry picked fix under PR 70633. Once that is merged, a rebase can be done for this commit.

File: pkg/kubelet/dockershim/network/plugins.go
Comment (neolit123) [inline pkg/kubelet/dockershim/network/plugins.go]: small whitespace issue.

File: cmd/kubeadm/app/phases/etcd/local.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/etcd/local.go]: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/etcd/local_test.go
would need some updates too?

File: cmd/kubeadm/app/phases/upgrade/staticpods.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/upgrade/staticpods.go]: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/upgrade/staticpods_test.go
would need dual stack coverage too.

**Review by neolit123 — COMMENTED** [review]
@pmichali thank you for working on this change.
added some minor comments.

this would need a release note instead of NONE.
@kubernetes/sig-cluster-lifecycle-pr-reviews 
/priority important-longterm

File: pkg/kubelet/dockershim/network/plugins.go
Comment (pmichali) [reply pkg/kubelet/dockershim/network/plugins.go]: Will address when I rebase.

**Comment by pmichali** [comment]
Updated commit, removing dependent changes from PR #70633, which has since merged. This is ready for review.

Two questions...

This uses the DNS_SVC_IP environment variable, which was introduced with IPv6 only capability previously, to determine if the IPv4 or IPv6 address is used for pod. If not specified, it will default back to trying to find the IPv4 address on the pod first, and then trying to find the IPv6 address.  Would the requirement of this environment variable constitute a "user facing" change?

If not, how do I remove the release note label? I accidentally deleted the text that is used to trigger this label, and only have the text "NONE", and I don't remember what the syntax is supposed to be.

**Comment by pmichali** [comment]
Reviewers, PTAL, and let me know about the release note question I posed.  Thanks!

**Comment by pmichali** [comment]
@neolit123 Can you take a look at the latest changes? This is ready for review. Thanks!

File: pkg/kubelet/dockershim/network/plugins.go
Comment (neolit123) [inline pkg/kubelet/dockershim/network/plugins.go]: i think it make sense to indicate that the fall back here is to -4.

simplest way is to use a `goto` to a label before:
```
fallback:
	glog.V(3).Infof("Using IPv4.....
```
yet i see that `goto` is not that widely used in the project.

File: pkg/kubelet/dockershim/network/plugins.go
Comment (neolit123) [reply pkg/kubelet/dockershim/network/plugins.go]: this whitespace issue seems to be present still.

**Review by neolit123 — APPROVED** [review]
@pmichali thanks for the update.
LGTM mostly. added a couple of minor comments.

sig-node and the kubelet maintainers would do the final LGTM / approve.
/ok-to-test

**Comment by neolit123** [comment]
also please add a release note as per the PR template:
https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md

File: pkg/kubelet/dockershim/network/plugins.go
Comment (pmichali) [reply pkg/kubelet/dockershim/network/plugins.go]: I'm changing the wording of the message. Let me know if that is OK.

File: pkg/kubelet/dockershim/network/plugins.go
Comment (pmichali) [reply pkg/kubelet/dockershim/network/plugins.go]: Sorry, I missed it. Had spaces vs tab, so it looked fine in my editor. Will be in next patch.

**Review by pmichali — COMMENTED** [review]
Will upload newer patch today.

**Comment by pmichali** [comment]
@neolit123 Thanks for looking. I was wondering if a release note is needed for this change. See my post 5 days ago... I'm thinking not, but need guidance here.

**Comment by pmichali** [comment]
@freehan @dchen1107 Can you PTAL at this kubelet change?  Also, regarding my question above, is a release note needed? If so, I could use a bit of guidance on the wording. If not, how do I remote the release note label?

**Comment by pmichali** [comment]
@thockin  could you help getting some eyes on this. I'm running out of time to be able to work on getting this merged (taking on new position at work with limited upstream involvement). Thanks!

**Comment by internetionals** [comment]
Still an issue on 1.14

/reopen
/remove-lifecycle-rotten
/remove-lifecycle-sta

**Pull request closed without merging**
